Core/Git/CommandLineGitClient.cs

```diff
@@ -442,25 +444,66 @@ private string GetFirstAncestor ()
   private CommandLineResult ExecuteGitCommandWithOutput (string arguments)
   {
     _log.Debug("Executing git command with output: git '{Arguments}'.", arguments);
+    using var process = new Process();
 
-    var psi = new ProcessStartInfo("git", arguments)
-              {
-                  UseShellExecute = false,
-                  RedirectStandardOutput = true,
-                  RedirectStandardError = true
-              };
+    process.StartInfo.FileName = "git";
+    process.StartInfo.Arguments = arguments;
+    process.StartInfo.UseShellExecute = false;
+    process.StartInfo.RedirectStandardOutput = true;
+    process.StartInfo.RedirectStandardError = true;
 
-    using var command = Process.Start(psi);
-    var output = command!.StandardOutput.ReadToEnd();
-    command!.WaitForExit(c_gitProcessWaitTimeout);
+    StringBuilder output = new StringBuilder();
+    StringBuilder error = new StringBuilder();
 
-    _log.Information($"git output: {output}");
+    using (AutoResetEvent outputWaitHandle = new AutoResetEvent(false))
+    using (AutoResetEvent errorWaitHandle = new AutoResetEvent(false))
+    {
+      process.OutputDataReceived += (_, e) =>
+      {
+        if (e.Data == null)
+        {
+          outputWaitHandle.Set();
+        }
+        else
+        {
+          output.AppendLine(e.Data);
+        }
+      };
+      process.ErrorDataReceived += (_, e) =>
+      {
+        if (e.Data == null)
+        {
+          errorWaitHandle.Set();
+        }
+        else
+        {
+          error.AppendLine(e.Data);
+        }
+      };
+
+      process.Start();
+
+      var timeout = 30_000;
+
+      process.BeginOutputReadLine();
+      process.BeginErrorReadLine();
+      if (process.WaitForExit(timeout) &&
+          outputWaitHandle.WaitOne(timeout) &&
+          errorWaitHandle.WaitOne(timeout))
+      {
+        _log.Information($"git output: {output}");
 
-    var success = command.ExitCode == 0;
-    if (!success)
-      return new CommandLineResult(success, command.StandardError.ReadToEnd());
+        var success = process.ExitCode == 0;
+        if (!success)
+          return new CommandLineResult(success, error.ToString());
 
-    return new CommandLineResult(success, output);
+        return new CommandLineResult(success, output.ToString().Trim());
+      }
+      else
+      {
+        throw new InvalidOperationException($"Could not successfully call 'git' with arguments '{arguments}'.");
```

Comment:
Exception instead of `CLR`: The program can't really continue when any of these methods fail, so we would have to kill it anyways when these errors happen.

Reply:
Timespan adds weird conversion issues when getting back the milliseconds, and the formatting is not really required when we just want to say ms or seconds anyways, so I'd like to keep it as ints with  ms inside :)

Reply:
Exception: When there is an error (non-zero exit code), we return an error result to the caller. But when there is a timeout, we throw an exception. I don't understand yet why our response to the caller is so different in these two scenarios? Maybe the caller can retry in the first scenario, but not in the second? But if so, should we decide this within the method?

Timespan: There is an overload `Process.WaitForExit(Timespan timeout)` and also `WaitHandle.WaitOne(TimeSpan timeout)` - we should not need to convert back to milliseconds. But if there is a problem, we can stick to ms.

Reply:
The timeout only occurs when something went really, really wrong and should never occur, so all bets are off. 
The non-zero exit code means that something unexpected was required which could not be fulfilled, like trying to check out a branch which does not exist and is a normal part of how the application should handle things. In most cases, this creates an exception at the calling site anyways. 

As for the timeout thing:
![image](https://github.com/re-motion/ReleaseProcessAutomation/assets/58393218/39df425a-d65c-4374-b729-743d24e41088)

Reply:
@Frazerus & @martin-herwich  the Timespan overload is availble in .NET 7 +

Reply:
Might as well upgrade ;)